fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,719 @@
+        let params = IPCodecParams::deserialize(&mut &data_footer[..])?;
+        let decompressor = CompactSpaceDecompressor { data, params };
+
+        Ok(decompressor)
+    }
+
+    /// Converting to compact space for the decompressor is more complex, since we may get values
+    /// which are outside the compact space. e.g. if we map
+    /// 1000 => 5
+    /// 2000 => 6
+    ///
+    /// and we want a mapping for 1005, there is no equivalent compact space. We instead return an
+    /// error with the index of the next range.
+    fn to_compact(&self, value: u128) -> Result<u64, usize> {
+        self.params.compact_space.to_compact(value)
+    }
+
+    fn compact_to_u128(&self, compact: u64) -> u128 {
+        self.params.compact_space.unpack(compact)
+    }
+
+    /// Comparing on compact space: 1.08 GElements/s, which equals a throughput of 17,3 Gb/s
+    /// (based on u128 = 16byte)
+    ///
+    /// Comparing on original space: .06 GElements/s (not completely optimized)
+    pub fn get_range(&self, range: RangeInclusive<u128>) -> Vec<u64> {
+        if range.start() > range.end() {
+            return Vec::new();
+        }
+        let from_value = *range.start();
+        let to_value = *range.end();
+        assert!(to_value >= from_value);
+        let compact_from = self.to_compact(from_value);
+        let compact_to = self.to_compact(to_value);
+
+        // Quick return, if both ranges fall into the same non-mapped space, the range can't cover
+        // any values, so we can early exit
+        match (compact_to, compact_from) {
+            (Err(pos1), Err(pos2)) if pos1 == pos2 => return Vec::new(),
+            _ => {}
+        }
+
+        let compact_from = compact_from.unwrap_or_else(|pos| {
+            // Correctness: Out of bounds, if this value is Err(last_index + 1), we early exit,
+            // since the to_value also mapps into the same non-mapped space
+            let range_mapping = self.params.compact_space.get_range_mapping(pos);
+            range_mapping.compact_start
+        });
+        // If there is no compact space, we go to the closest upperbound compact space
+        let compact_to = compact_to.unwrap_or_else(|pos| {
+            // Correctness: Overflow, if this value is Err(0), we early exit,
+            // since the from_value also mapps into the same non-mapped space
+
+            // Get end of previous range
+            let pos = pos - 1;
+            let range_mapping = self.params.compact_space.get_range_mapping(pos);
+            range_mapping.compact_start + range_mapping.range_length()
+        });
+
+        let range = compact_from..=compact_to;
+        let mut positions = Vec::new();
+
+        let step_size = 4;
+        let cutoff = self.params.num_vals - self.params.num_vals % step_size;
+
+        let mut check_add = |idx, val| {
+            if range.contains(&val) && val != self.params.null_value_compact_space {
```

Comment:
Why is it needed to check for the null value!?

If we put the null value in compact at the extremity, 0 or MAX+1 it should not be necessary to do that check,

Reply:
yes that's better and will be faster. For reference I'll copy the discord message here

> In the old version, the user had to replace None with the null value the compressor chose, because the API was taking Vec<u128>. So the null value was not just a compact space thing. But this api sucked, in the new version we accept Vec<Option<u128>>. So this null value dance can actually be changed, because null is now only a  compact space thing. This should simplify things